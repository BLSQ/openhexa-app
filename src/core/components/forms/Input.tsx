import { InputHTMLAttributes, forwardRef, ReactNode } from "react";
import clsx from "clsx";

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  error?: string | null | false | true;
  trailingIcon?: ReactNode;
}

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const {
    type,
    name,
    error,
    className,
    trailingIcon,
    value = "",
    ...delegated
  } = props;

  const inputClassName = clsx(
    "w-full form-input appearance-none relative block",
    "px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm",
    "disabled:opacity-50 disabled:cursor-not-allowed",
    error
      ? "border-red-300  placeholder-red-300  text-red-900  focus:ring-red-500  focus:border-red-500"
      : "border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500",
    trailingIcon && "pr-4",
    className
  );

  return (
    <div className="group relative">
      <input
        id={name}
        name={name}
        type={type ?? "text"}
        aria-invalid={Boolean(error)}
        aria-describedby={`${name}-description`}
        className={inputClassName}
        value={value}
        {...delegated}
        ref={ref}
      />
      {trailingIcon && (
        <div className="absolute inset-y-0 right-0 inline-flex items-center justify-center pr-2.5">
          {trailingIcon}
        </div>
      )}
    </div>
  );
});

Input.displayName = "Input";

export default Input;
