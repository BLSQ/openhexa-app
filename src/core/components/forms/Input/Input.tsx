import { InputHTMLAttributes, forwardRef, ReactNode } from "react";
import clsx from "clsx";

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  error?: string | null | false | true;
  trailingIcon?: ReactNode;
  leading?: ReactNode;
  fullWidth?: boolean;
}

const Input = forwardRef<HTMLInputElement, InputProps>((props, ref) => {
  const {
    type,
    name,
    error,
    defaultValue,
    className,
    trailingIcon,
    leading = null,
    value,
    fullWidth = false,
    ...delegated
  } = props;

  const inputClassName = clsx(
    "form-input appearance-none relative block",
    "px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm",
    "disabled:opacity-50 disabled:cursor-not-allowed",
    error
      ? "border-red-300 placeholder-red-300  text-red-900  focus:ring-red-500  focus:border-red-500"
      : "border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500",
    trailingIcon && "pr-4",
    leading && "pl-10",
    fullWidth && "w-full",
    className,
  );

  return (
    <div className={clsx("group relative", fullWidth && "w-full")}>
      <input
        id={name}
        name={name}
        type={type ?? "text"}
        aria-invalid={Boolean(error)}
        aria-describedby={`${name}-description`}
        className={inputClassName}
        value={defaultValue ? undefined : value ?? ""}
        defaultValue={defaultValue}
        {...delegated}
        ref={ref}
      />
      {leading && (
        <div className="absolute inset-y-0 left-0 z-30 inline-flex items-center justify-center pl-2.5">
          {leading}
        </div>
      )}
      {trailingIcon && (
        <div className="absolute inset-y-0 right-0 z-10 inline-flex items-center justify-center pr-2.5">
          {trailingIcon}
        </div>
      )}
    </div>
  );
});

Input.displayName = "Input";

export default Input;
