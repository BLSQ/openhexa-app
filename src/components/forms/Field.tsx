import Label from "./Label";
import { ReactElement, ReactNode } from "react";
import Input, { InputProps } from "./Input";
import clsx from "clsx";
import Tooltip from "components/Tooltip";
import { InformationCircleIcon } from "@heroicons/react/solid";

interface CommonProps {
  label?: string;
  error?: string | null | false;
  required?: boolean;
  description?: string;
  name: string;
  help?: ReactNode;
  className?: string;
  labelColor?: string;
  errorColor?: string;
}

type CustomField = CommonProps & {
  children: ReactElement | ReactElement[];
};

type InputField = CommonProps & InputProps;

const Field = (props: CustomField | InputField) => {
  const {
    error,
    description,
    label,
    required,
    name,
    help,
    className,
    labelColor,
    errorColor = "text-red-600",
    ...delegated
  } = props;

  let children;
  if (props.children) {
    children = props.children;
  } else {
    children = (
      <Input name={name} error={error} required={required} {...delegated} />
    );
  }

  return (
    <div className={className}>
      {label && (
        <div className="flex justify-between mb-1">
          <Label color={labelColor} htmlFor={name} className="flex">
            {label}
            {help && (
              <Tooltip
                placement="top"
                renderTrigger={(ref) => (
                  <span ref={ref}>
                    <InformationCircleIcon className="ml-1 h-3 w-3 cursor-pointer" />
                  </span>
                )}
                label={help}
              />
            )}
          </Label>
          {!required && <span className="text-sm text-gray-400">Optional</span>}
        </div>
      )}
      {children}
      {description && (
        <p className="mt-2 text-sm text-gray-500">{description}</p>
      )}
      {error && <p className={clsx("mt-2 text-sm", errorColor)}>{error}</p>}
    </div>
  );
};

export default Field;
