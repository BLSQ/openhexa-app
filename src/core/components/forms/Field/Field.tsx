import Label from "../Label";
import { ReactElement, ReactNode } from "react";
import Input from "../Input";
import clsx from "clsx";
import Tooltip from "core/components/Tooltip";
import { InformationCircleIcon } from "@heroicons/react/20/solid";

interface CommonProps {
  label?: string;
  error?: string | null | false;
  required?: boolean;
  description?: ReactNode | string;
  showOptional?: boolean;
  name: string;
  help?: ReactNode;
  className?: string;
  labelColor?: string;
  errorColor?: string;
}

type CustomField = CommonProps & {
  children: ReactElement | ReactElement[];
};

type InputField = CommonProps & React.ComponentProps<typeof Input>;

const Field = (props: CustomField | InputField) => {
  const {
    error,
    description,
    label,
    required,
    name,
    help,
    showOptional = true,
    className,
    labelColor,
    errorColor = "text-red-600",
    ...delegated
  } = props;

  let children;
  if (props.children) {
    children = props.children;
  } else {
    children = (
      <Input name={name} error={error} required={required} {...delegated} />
    );
  }

  return (
    <div className={className}>
      {label && (
        <div className="mb-1 flex justify-between">
          <Label htmlFor={name} className={clsx("flex", labelColor)}>
            {label}
            {help && (
              <Tooltip
                placement="top"
                renderTrigger={(ref) => (
                  <span ref={ref} data-testid="help">
                    <InformationCircleIcon className="ml-1 h-3 w-3 cursor-pointer" />
                  </span>
                )}
                label={help}
              />
            )}
          </Label>
          {!required && showOptional && (
            <span className="text-sm text-gray-400">Optional</span>
          )}
        </div>
      )}
      {children}
      {description && (
        <p className="mt-2 text-sm text-gray-500">{description}</p>
      )}
      {error && <p className={clsx("mt-2 text-sm", errorColor)}>{error}</p>}
    </div>
  );
};

export default Field;
