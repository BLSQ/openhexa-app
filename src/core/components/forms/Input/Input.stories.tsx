import { Story } from "@ladle/react";
import Input from "./Input";

export const InputStory: Story = (props) => <Input {...props} />;

InputStory.storyName = "default";
InputStory.args = {};
InputStory.argTypes = {};

const defaults = { title: "UI / Forms / Input" };

export default defaults;
