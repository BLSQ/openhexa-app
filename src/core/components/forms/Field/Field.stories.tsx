import { Story, action } from "@ladle/react";

import Field from "./Field";
import Input from "../Input";
import Checkbox from "../Checkbox";

export const FieldStory: Story<React.ComponentProps<typeof Field>> = (
  props,
) => (
  <div className="space-y-4">
    <Field {...props}>
      <Input value="" placeholder="This one is dynamic" onChange={() => {}} />
    </Field>

    <Field
      name="number"
      label={"number"}
      required
      help="Help text"
      type="number"
      value={42}
      onChange={() => {}}
    />
  </div>
);
FieldStory.storyName = "default";
FieldStory.args = {
  required: false,
  label: "Field label",
  description: "",
  name: "",
  className: "",
  labelColor: "",
  errorColor: "",
  error: "",
};

FieldStory.argTypes = {};

const defaults = {
  title: "UI / Forms / Field",
};

export default defaults;
