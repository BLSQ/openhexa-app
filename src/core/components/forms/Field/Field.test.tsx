import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Input from "../Input";
import Field from "./Field";

describe("Field", () => {
  it("renders", async () => {
    const { container } = render(
      <Field name="field" label={"Field"}>
        <Input name="field" value="" onChange={() => {}} />
      </Field>,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders a input if no child is given", async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Field
        name="field"
        label={"Field"}
        onChange={onChange}
        value={"Value"}
      />,
    );

    expect(screen.getByDisplayValue("Value")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("displays a question mark if help is provided", () => {
    render(
      <Field name="field" label={"Field"} help="Help">
        <span>Hey</span>
      </Field>,
    );
    expect(screen.getByTestId("help")).toBeInTheDocument();
  });
});
