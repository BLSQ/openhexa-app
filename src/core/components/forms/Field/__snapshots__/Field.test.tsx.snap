// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders 1`] = `
<div>
  <div>
    <div
      class="mb-1 flex justify-between"
    >
      <label
        class="block text-sm font-medium text-gray-600 flex"
        for="field"
      >
        Field
      </label>
      <span
        class="text-sm text-gray-400"
      >
        Optional
      </span>
    </div>
    <div
      class="group relative"
    >
      <input
        aria-describedby="field-description"
        aria-invalid="false"
        class="form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
        id="field"
        name="field"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Field renders a input if no child is given 1`] = `
<div>
  <div>
    <div
      class="mb-1 flex justify-between"
    >
      <label
        class="block text-sm font-medium text-gray-600 flex"
        for="field"
      >
        Field
      </label>
      <span
        class="text-sm text-gray-400"
      >
        Optional
      </span>
    </div>
    <div
      class="group relative"
    >
      <input
        aria-describedby="field-description"
        aria-invalid="false"
        class="form-input appearance-none relative block px-3 py-2 border rounded-md focus:outline-none focus:z-10 sm:text-sm disabled:opacity-50 disabled:cursor-not-allowed border-gray-300 placeholder-gray-500 text-gray-900 focus:ring-blue-500 focus:border-blue-500"
        id="field"
        name="field"
        type="text"
        value="Value"
      />
    </div>
  </div>
</div>
`;
